File: 2018_summer/land.py
def solution(land, P, Q):

    land_1d = []
    for row in land:
        land_1d += row
    land_1d.sort()
    maxh = land_1d[-1]
    minh = land_1d[0]    

    def func(t):
        return sum([(t-c)*P if t-c >= 0 else (c-t)*Q for c in land_1d])

    answer = []
    import heapq

    left = minh
    right = maxh
    x = (left+right)//2
    while left < right:
        if func(x) < func(x+1) :
            right = x
        else :
            left = x+1
        x = (left+right)//2
    heapq.heappush(answer, func(x)) 

    return heapq.heappop(answer)
land = [[4, 4, 3], [3, 2, 2], [ 2, 1, 0 ]]
n = solution(land, 5, 3)                    
print(n)

File: 2018_summer/land.cpp
#include <vector>
#include <algorithm>

using namespace std;

int n ;
int myP, myQ;
vector<int> myland;


long long func(long target){
    long long result = 0;
    for(int i = 0; i < n*n; i++){
         result += (long long)(target - myland[i]>= 0 ? (target-myland[i])*myP : (myland[i]-target)*myQ);
    }
    return result;
}
long long solution(vector<vector<int> > land, int P, int Q)
{
    myP = P;
    myQ = Q;
    n = land.size();
    vector<long long> answer;
    for(int i = 0; i < n; i++)
        for(int j =0; j<n; j++)
            myland.push_back(land[i][j]);
    sort(myland.begin(), myland.end());
    int maxh = myland[n*n-1];
    int minh = myland[0];

    int left = minh;
    int right = maxh;
    int x = (left+right)/2;
    while(left < right){

        if(func(x) < func(x+1)){
            right = x ;
        }else {
            left = x+1;
        }
        x = (left+ right)/2;
    }
    return func(x);


}